src/app/util/af-types.h

```diff
@@ -423,14 +425,16 @@ typedef struct
  *   and pass a pointer to that location around during
  *   command processing
  */
-typedef struct
+struct EmberAfClusterCommand
 {
+    chip::NodeId SourceNodeId() const { return source->GetSecureSession().GetPeerNodeId(); }
+
     /**
      * APS frame for the incoming message
      */
     EmberApsFrame * apsFrame;
     EmberIncomingMessageType type;
-    chip::NodeId source;
+    chip::Messaging::ExchangeContext * source;
```

Comment:
The exchange on the receiving  side is closed immediately within unsolicited message handler? 

void OnMessageReceived(Messaging::ExchangeContext * exchangeContext, const PacketHeader & packetHeader,
                           const PayloadHeader & payloadHeader, System::PacketBufferHandle buffer) override
    {
.....
        {
            HandleDataModelMessage(exchangeContext, std::move(buffer));
        }

    exit:
        exchangeContext->Close();
}

How this passed-in exchange within EmberAfClusterCommand is used? how can we make sure the passed-in exchange is not dangling pointer before use?

Reply:
It's used to synchronously, before `HandleDataModelMessage` returns, send a response.

Reply:
So the API contract for now is that `HandleDataModelMessage` must respond sync.  If that changes, we will need to condition the `Close` here on `HandleDataModelMessage` not being reached and hand off closing to `HandleDataModelMessage`.

Reply:
Acked

Reply:
The exchange needs to be closed at the end of the 'conversation', this syntax is defined within the protocol since it has the context to know which message mark the end of the conversation.

For TempZCL protocol without a protocol syntax defined, I feel it is more safe to use a dedicated exchange to send  request and response message separately.

Reply:
I'm not sure what you mean by "the response needs to be sent from different bindings".

Right now, the contract is that `emberAfPreCommandReceivedCallback` would need to send a response, if any, before returning.

Again, we could change this contract.  I don't think we should do it in this PR.

Reply:
> For TempZCL protocol without a protocol syntax defined, I feel it is more safe to use a dedicated exchange to send request and response message separately.

There are a few problems with that:

1. It leads to problems correlating responses with requests.
2. It causes our existing tests to not exercise various exchange-related code (like CRMP).  Note that in the process of working on this PR I ended up finding various controller and CRMP bugs that were due precisely due to that functionality not being exercised in existing tests.

In general, everything in TempZCL is currently of the form (1 request, 1 response).  The one exception I can think of is whether we ever send Default Response from the _initiator_ side (in response to a non-status response).  But since the responder side ignores those responses anyway, it doesn't matter too much that it closes its exchange before it receives them.

Reply:
Yes, I think we can still benefit more from this change if we can guarantee HandleDataModelMessage must handle the respond synchronously.

CRMP  is disabled since separate exchange is used for request and response, should we enable CRMP now?

Reply:
This PR enables CRMP on the DeviceController side and enables it for responses to an incoming message on the data model side.  The one place we are still disabling it is for data-model-initiated sends to a specific node (e.g. attribute reporting).